pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,195 @@
+            jac = qml.math.tensordot(qjac, cjac, [[-1], [0]])
+            jac = qml.math.tensordot(jac, cjac, [[-2], [0]])
+            return qml.math.swapaxes(jac, -1, -2)
+
+        return jacobian_wrapper
+
+
+@hessian_transform
+def param_shift_hessian(tape):
+    """Generate parameter-shift tapes and postprocessing method
+    to directly compute the hessian."""
+
+    # perform gradient method validation
+    if any(m.return_type is qml.operation.State for m in tape.measurements):
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    _gradient_analysis(tape)
+    gradient_tapes = []
+    gradient_coeffs = []
+    shapes = []
+    dim = len(tape.trainable_params)
+
+    if not tape.trainable_params:
+        return gradient_tapes, lambda _: np.zeros([tape.output_dim, len(tape.trainable_params)])
+
+    # The Hessian for a 2-term parameter-shift rule can be expressed via the following recipes.
+    # Off-diagonal elements of the Hessian require shifts to two different parameter indices.
+    # A recipe can thus be expressed via the tape patterns: (dummy values for ndarray creation)
+    #       [[coeff, dummy], [mult, dummy], [shift1, shift2]]
+    # Each corresponding to one term in the parameter-shift formula:
+    #       didj f(x) = coeff * f(mult*x + shift1*ei +shift2*ej) + ...
+    diag_recipe = [
+        [[ 0.5], [1], [np.pi]],
+        [[-0.5], [1], [    0]]
+    ]
+    off_diag_recipe = [
+        [[ 0.25, 1], [1, 1], [ np.pi/2,  np.pi/2]],
+        [[-0.25, 1], [1, 1], [-np.pi/2,  np.pi/2]],
+        [[-0.25, 1], [1, 1], [ np.pi/2, -np.pi/2]],
+        [[ 0.25, 1], [1, 1], [-np.pi/2, -np.pi/2]]
+    ]
+
+    # for now assume all operations support the 2-term parameter shift rule
+    for idx in product(range(len(tape.trainable_params)), repeat=2):
+
+        recipe = diag_recipe if idx[0] == idx[1] else off_diag_recipe
+        coeffs, multipliers, shifts = _process_gradient_recipe(recipe)
+
+        # generate the gradient tapes
+        gradient_coeffs.append(coeffs)
+        g_tapes = generate_multishifted_tapes(tape, idx, shifts)
+
+        gradient_tapes.extend(g_tapes)
+        shapes.append((idx, len(g_tapes)))
+
+    def processing_fn(results):
+        grads = []
+        start = 0
+
+        for i, (idx, s) in enumerate(shapes):
+            res = results[start : start + s]
+            start = start + s
+
+            # compute the linear combination of results and coefficients
+            res = qml.math.squeeze(qml.math.stack(res))
+            g = qml.math.tensordot(res, qml.math.convert_like(gradient_coeffs[i], res), [[0], [0]])
+
+            for j, elem in enumerate(qml.math.atleast_1d(g)):
+                if j < len(grads):
+                    grads[j][idx[0], idx[1]] = elem
+                else:
+                    hessian = qml.math.empty((dim,dim))
```

Comment:
You may need to be careful here! Qnode's can have vector-valued output (e.g., `return qml.probs`.
```suggestion
                    hessian = qml.math.empty((dim, dim) + elem.shape)
```

Reply:
I think that should be accounted for, because here I just create the "innermost" Hessian. If the QNode is vector valued, `g` is too and I generate multiple of these "inner" Hessians stacked together, where the outer dimension represents the shape of the QNode output.
This is apparently the shape created by Jacobian(Jacobian(QNode)). However, I've not tested a case where the QNode output is higher dimension than 1D (as you've pointed out in a comment above), so I'll have to take a look at this.

Reply:
eb18240 should fix this for any QNode output dimensions. It was indeed only handling 0D and 1D cases correctly. Thanks :)

Reply:
Awesome!